### PR TITLE
[MWAR-220] filter overlay jars conflicting with managed dependencies

### LINKED ISSUE #389: [MWAR-220] dependencyManagement && overlay + transitive dependency = multiple versions of same dependency in WEB-INF/lib
**[jieryn](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jieryn)** opened **[MWAR-220](https://issues.apache.org/jira/browse/MWAR-220?redirect=false)** and commented

With dependencyManagement section defined I get multiple versions of spring libraries in target/cas-*/WEB-INF/lib after mvn package. If I remove the dependencyManagement section then only one version of the spring libraries are found.

There seems to be a problem with m-war-p and overlays and dependencyManagement not using the maven GAV resolver in a proper way.. I attached a simple pom which exhibits the problem, just run mvn package to verify.

---

**Attachments:**
- [pom.xml](https://issues.apache.org/jira/secure/attachment/12720634/pom.xml) (_5.99 kB_)

3 votes, 6 watchers

### PR DESCRIPTION
**Pull request by elharo**
Fixes #389

When a project uses dependencyManagement to pin a specific version of a library, any jar in a WAR overlay's `WEB-INF/lib` with the same artifactId but a different version will now be removed during overlay processing.

### How it works
In `OverlayPackagingTask.performPackaging()`, after unpacking the overlay and before copying files, the new `filterConflictingDependencyJars()` method scans the overlay's `WEB-INF/lib` directory. For each jar, it extracts the artifactId from the filename (following the `artifactId-version(-classifier)?.jar` convention). If the artifactId matches a non-optional jar artifact from the project's dependency tree, the overlay jar is removed — the managed version will be provided by the project's own dependency resolution.

### Additional changes
- Updated the MWAR-389 IT test (`verify.bsh`) to remove print statements for the passing case per review feedback.

The IT test in PR #638 documents this issue and now passes with this fix applied.

### CODE REVIEW

File: src/it/MWAR-389/verify.bsh [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+if ( overlayHasManaged )
+{
+    System.err.println( "Overlay WAR unexpectedly contains plexus-utils-" + managedVersion + ".jar. "
+        + "Overlay should only have version 3.0.24." );
+    return false;
+}
+
+System.out.println( "Overlay WAR correctly contains only plexus-utils-3.0.24.jar." );
```
Comment (elharo) [inline src/it/MWAR-389/verify.bsh]: remove this; passing tests generate no output

File: src/it/MWAR-389/verify.bsh [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+if ( !mainHasManaged )
+{
+    System.err.println( "Main WAR missing plexus-utils-" + managedVersion + ".jar in WEB-INF/lib." );
+    return false;
+}
+
+if ( mainHasOld )
+{
+    // Both versions found -- issue STILL EXISTS
+    System.out.println( "ISSUE #389 (MWAR-220) STILL EXISTS:" );
```
Comment (elharo) [inline src/it/MWAR-389/verify.bsh]: remove this; passing tests generate no output

File: src/it/MWAR-389/verify.bsh [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    System.out.println( "  Both plexus-utils versions found in main WAR WEB-INF/lib:" );
+    System.out.println( "    - plexus-utils-3.0.24.jar (from overlay unpacking)" );
```
Comment (elharo) [inline src/it/MWAR-389/verify.bsh]: remove this; passing tests generate no output

File: src/it/MWAR-389/verify.bsh [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+{
+    System.err.println( "Main WAR missing plexus-utils-" + managedVersion + ".jar in WEB-INF/lib." );
+    return false;
+}
+
+if ( mainHasOld )
+{
+    // Both versions found -- issue STILL EXISTS
+    System.out.println( "ISSUE #389 (MWAR-220) STILL EXISTS:" );
+    System.out.println( "  Both plexus-utils versions found in main WAR WEB-INF/lib:" );
```
Comment (elharo) [inline src/it/MWAR-389/verify.bsh]: remove this; passing tests generate no output

File: src/it/MWAR-389/verify.bsh [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    System.out.println( "    - plexus-utils-3.0.24.jar (from overlay unpacking)" );
+    System.out.println( "    - plexus-utils-" + managedVersion + ".jar (from dependencyManagement)" );
+    System.out.println( "  dependencyManagement does not filter overlay-provided jars." );
```
Comment (elharo) [inline src/it/MWAR-389/verify.bsh]: remove this; passing tests generate no output

File: src/it/MWAR-389/verify.bsh [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    return false;
+}
+
+if ( mainHasOld )
+{
+    // Both versions found -- issue STILL EXISTS
+    System.out.println( "ISSUE #389 (MWAR-220) STILL EXISTS:" );
+    System.out.println( "  Both plexus-utils versions found in main WAR WEB-INF/lib:" );
+    System.out.println( "    - plexus-utils-3.0.24.jar (from overlay unpacking)" );
+    System.out.println( "    - plexus-utils-" + managedVersion + ".jar (from dependencyManagement)" );
```
Comment (elharo) [inline src/it/MWAR-389/verify.bsh]: remove this; passing tests generate no output

File: src/it/MWAR-389/verify.bsh [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    System.out.println( "  dependencyManagement does not filter overlay-provided jars." );
+    return false;
+}
+else
+{
+    // Only managed version -- issue has been fixed
+    System.out.println( "ISSUE #389 (MWAR-220) APPEARS FIXED:" );
```
Comment (elharo) [inline src/it/MWAR-389/verify.bsh]: remove this; passing tests generate no output

File: src/it/MWAR-389/verify.bsh [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    System.out.println( "  Only plexus-utils-" + managedVersion + ".jar found in WEB-INF/lib." );
```
Comment (elharo) [inline src/it/MWAR-389/verify.bsh]: remove this; passing tests generate no output

File: src/it/MWAR-389/verify.bsh [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    System.out.println( "  The overlay's plexus-utils-3.0.24.jar was correctly filtered out." );
```
Comment (elharo) [inline src/it/MWAR-389/verify.bsh]: remove this; passing tests generate no output

File: src/main/java/org/apache/maven/plugins/war/packaging/OverlayPackagingTask.java [hunk 10]
```diff
@@ -127,4 +137,96 @@ protected File getOverlayTempDirectory(WarPackagingContext context, Overlay over
...
+        for (File overlayJar : overlayJars) {
+            String jarName = overlayJar.getName();
+            String artifactId = extractArtifactId(jarName);
+            if (artifactId != null && projectArtifactIds.contains(artifactId)) {
+                context.getLog()
+                        .debug("Excluding dependency [" + jarName + "] from overlay [" + overlay.getId()
+                                + "]; project runtime dependencies already include a version");
+                conflicting.add("WEB-INF/lib/" + jarName);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/main/java/org/apache/maven/plugins/war/packaging/OverlayPackagingTask.java:201]: `overlay` is a field of `OverlayPackagingTask` (line 46: `private final Overlay overlay;`). Since `computeConflictingJars` is an instance method, `this.overlay` is accessible in scope. The code compiles and runs correctly.

File: src/main/java/org/apache/maven/plugins/war/packaging/OverlayPackagingTask.java [hunk 11]
```diff
@@ -127,4 +137,96 @@ protected File getOverlayTempDirectory(WarPackagingContext context, Overlay over
...
+            return Collections.emptySet();
+        }
+
+        ScopeArtifactFilter filter = new ScopeArtifactFilter(Artifact.SCOPE_RUNTIME);
+        Set<String> projectArtifactIds = new HashSet<>();
+        for (Artifact artifact : context.getProject().getArtifacts()) {
+            if (!artifact.isOptional() && filter.include(artifact) && "jar".equals(artifact.getType())) {
+                projectArtifactIds.add(artifact.getArtifactId());
+            }
+        }
```
[the remark this answers is not in the document]
Comment (elharo) [reply src/main/java/org/apache/maven/plugins/war/packaging/OverlayPackagingTask.java:173]: Addressed in commit fc54f70. Added `getJarGroupId()` which reads `META-INF/maven/**/pom.properties` from each overlay jar to obtain the groupId for precise `groupId:artifactId` matching. Falls back to artifactId-only matching when the metadata is absent.

**Review by ascheman — CHANGES_REQUESTED** [review]
Thanks @elharo — the overlay-conflict filtering is a useful fix and the structure is sound (scope/optional/type filter, exclude-not-mutate). A few things before it's ready:

**1. Rebase onto current master.** The branch is 7 commits behind and predates #634 ([MWAR-471], permission normalization). Worth rebasing here specifically because:
- `OverlayPackagingTask extends AbstractWarPackagingTask`, whose `copyFile()` now normalizes permissions (#634) — so after the rebase the kept overlay jars get the same treatment, keeping behavior consistent.
- #634 also added a `TestWarPackagingContext` test double (+ `AbstractWarPackagingTaskTest`) — a ready harness to unit-test this PR (point 3).

**2. Match on the jar's own coordinates from `pom.properties` when present.** `getJarGroupId()` already opens the jar and reads `META-INF/maven/**/pom.properties`, but returns only the `groupId` — while the `artifactId` is separately (and heuristically) parsed from the filename. Read the full **groupId + artifactId (+ version)** from that same `pom.properties`: it's authoritative and free (you're already in the file), gives an exact `groupId:artifactId` match, and sidesteps the filename guess entirely for Maven-built jars (the common case). The filename heuristic then only matters as a **fallback** for jars without that metadata.
- For that fallback, `extractArtifactId()` also scans the wrong direction: right-to-left finds the *last* `-<digit>`, so `foo-1.0-1.jar` → `foo-1.0` (should be `foo`), and timestamped snapshots mis-parse. Left-to-right (first `-<digit>`) is correct and still handles `commons-lang3-3.20.0.jar`.

**3. Broaden test coverage — and split it by layer:**

**Unit tests (new)** — the pure logic is where most coverage belongs, and today it's `private` and untested. Make the extraction/matching package-private (or a small helper) and, reusing the `TestWarPackagingContext` double added in #634, cover:
- *Parsing* (`extractArtifactId`), good/bad/edge: plain, digit-in-artifactId (`commons-lang3-3.20.0`), classifier, SNAPSHOT, `foo-1.0-1.jar`, timestamped snapshot, no-version, non-`.jar`, `null`.
- *Matching* (`computeConflictingJars`): a jar *with* `pom.properties` whose GAV matches a managed dependency → excluded (locks in point 2); a **non-conflicting** jar → **preserved**; a same-artifactId/**different-groupId** jar → **not** removed (false-positive guard); the no-`pom.properties` **fallback**; optional/provided project deps ignored.

**Integration test** — the existing `MWAR-389` IT already covers the positive end-to-end path (managed version wins, `3.0.24` dropped, build fails otherwise), so keep it. The one worthwhile IT *addition* is the **preservation** case: add a jar to the overlay that is *not* a project dependency and assert it survives in the final WAR. The remaining negative/edge cases belong in the unit tests above rather than in more ITs.

**4. Rename the IT `MWAR-389` → `gh-389`.** This is GH #389 (JIRA key MWAR-**220**, per the title), so `MWAR-389` mixes the JIRA prefix with the GH number; the repo already uses the `gh-<n>` convention (e.g. `gh-554`). Minor: `invoker.goals=clean install` → `clean verify` — it's a reactor, so the overlay resolves without installing to the repo.

Happy to re-review once it's rebased, matching on metadata coordinates, with the parser + negative-case tests and the rename.

**Review by ascheman — CHANGES_REQUESTED** [review]
Thanks for the thorough rework, Elliott — groupId-from-metadata, the dependencyManagement conflict IT, the rebase, and the `TestWarPackagingContext` double are all real improvements, and CI is green across the matrix. Three things before I can approve:

**1. Read `artifactId` from the same `pom.properties`.** `getJarGroupId` already opens `META-INF/maven/**/pom.properties` — and that file always carries `artifactId` (and `version`) alongside `groupId`. Right now `artifactId` still comes from `extractArtifactId`'s filename heuristic, which scans right-to-left and mis-parses multi-hyphen versions (e.g. `foo-1.0-1.jar` → `foo-1.0`). Since you've already got the file open, read `artifactId` from it too and keep the filename parser only as a fallback for jars without Maven metadata. Then `groupId:artifactId` is fully authoritative for any Maven-built overlay jar.

**2. Unit tests for the matching.** The new `AbstractWarPackagingTaskTest` covers copy-permission behaviour but not the conflict-matching. Please add tests (via `TestWarPackagingContext`) for: metadata present (authoritative GAV), metadata absent (filename fallback), and the tricky names — `commons-lang3-3.20.0.jar`, `foo-1.0-1.jar`, classifier jars. That's the good/bad/edge coverage the parser needs.

**3. Rename the IT to the `gh-<n>` convention.** The IT is still `src/it/MWAR-389/`, but the repo has moved to GitHub-issue-numbered ITs (cf. `src/it/gh-554/`). Please rename it to `gh-389` (matching the tracking issue) so it's consistent with the current convention.

Minor, non-blocking: `invoker.goals=clean install` could be `clean verify` unless `install` is actually required.
